test/src/toga_test/app.py

```diff
@@ -0,0 +1,27 @@
+import pytest
+import tests
+
+import toga
+
+
+class TogaTest(toga.App):
+    def startup(self):
+        # For the "app" fixture.
+        global app
+        app = self
+
+        self.main_box = toga.Box()
+
+        self.main_window = toga.MainWindow(title=self.formal_name)
+        self.main_window.content = self.main_box
+        self.main_window.show()
+
+        self.add_background_task(test)
+
+
+def test(app):
+    pytest.main(["-vv", "--color=no"] + tests.__path__)
```

Comment:
This probably won't work on Android because its Python modules don't exist as separate files on the filesystem. As with the threading issue, it looks like this can be dealt with using pytest's hook system.

Reply:
I've added a chaquopy_extract_package function to the bottom of app.py which extracts the given package into individual files on the filesystem. This could be used in the testbed app as well.

A cleaner solution to this might be to resurrect a Chaquopy setting called [`extractPackages`](https://chaquo.com/chaquopy/doc/6.2/android.html#resource-files). This used to be required for any packages that expected to find its non-Python files on the filesystem, but it turned out there were so many packages like that, that I just made Chaquopy ignore the setting and extract data files automatically, but not Python files. 

That works for 99% of packages, but occasionally I get a user report of some code which expects Python files to exist on the filesystem as well.  So we could re-enable the `extractPackages` setting and make it mean "extract *all* files in this package, including Python files".

Reply:
In cases like this, the code probably expects to find files named `.py`, not `.pyc`. So for anything listed in `extractPackages`, Chaquopy should preserve the original .py files, which it normally omits if it succeeded in compiling them to .pyc format during the build.

For the moment, with pytest, we'll need to work around this by using the [`pyc`](https://chaquo.com/chaquopy/doc/current/android.html#android-bytecode) setting to disable build-time compilation, So the `python` section in app/build.gradle would look like this:
```
        python {
            version "3.8"
            pip {
                install "-r", "requirements.txt"
            }
            pyc {
                src false
            }
        }
```

To enable this, and the future use of `extractPackages`, and other user customizations, how about adding a pyproject.toml setting `build_gradle_extra_content`, that works the same way as [`dockerfile_extra_content`](https://github.com/beeware/briefcase-linux-appimage-template/pull/22)?

Gradle's brace syntax is really just shorthand for dotted notation. For example, the above block is equivalent to this:
```
        python {
            version "3.8"
            pip {
                install "-r", "requirements.txt"
            }
        }

        python {
            pyc {
                src false
            }
        }
```

Or this:
```
        python {
            version "3.8"
            pip {
                install "-r", "requirements.txt"
            }
        }

        python.pyc.src false
```

So adding the `build_gradle_extra_content` at the end of the file should be fine: it would still be able to customize any of the settings in the blocks above.

Reply:
Exposing Gradle customisation definitely seems like a reasonable thing to do - no matter what we put in the template, I'm sure *someone* will have other things they want to do. If it's as simple as adding strings of dotted paths to the bottom of a file, then it seems like a no-brainer.

I'd also be in favour of adding the `extractPackages` setting; it's got immediate utility for test suites, and I have to imagine there's a set of weird-edge case packages that will actually want to have their source files on disk for Reasons (tm).